File: mobile-engage/src/main/java/com/emarsys/mobileengage/device/DeviceInfoStartAction.java
package com.emarsys.mobileengage.device;

import android.app.Activity;

import com.emarsys.core.activity.ActivityLifecycleAction;
import com.emarsys.core.device.DeviceInfo;
import com.emarsys.core.storage.Storage;
import com.emarsys.core.util.Assert;
import com.emarsys.mobileengage.client.ClientServiceInternal;

public class DeviceInfoStartAction implements ActivityLifecycleAction {

    private final DeviceInfo deviceInfo;
    private final ClientServiceInternal clientInternal;
    private final Storage<String> deviceInfoPayloadStorage;

    public DeviceInfoStartAction(ClientServiceInternal clientInternal, Storage<String> deviceInfoPayloadStorage, DeviceInfo deviceInfo) {
        Assert.notNull(clientInternal, "ClientInternal must not be null!");
        Assert.notNull(deviceInfoPayloadStorage, "DeviceInfoPayloadStorage must not be null!");
        Assert.notNull(deviceInfo, "DeviceInfo must not be null!");

        this.clientInternal = clientInternal;
        this.deviceInfoPayloadStorage = deviceInfoPayloadStorage;
        this.deviceInfo = deviceInfo;
    }

    @Override
    public void execute(Activity activity) {
        if (deviceInfoPayloadStorage.get() == null || !deviceInfoPayloadStorage.get().equals(deviceInfo.getDeviceInfoPayload())) {
            clientInternal.trackDeviceInfo();
        }
    }
}

File: mobile-engage/src/main/java/com/emarsys/mobileengage/device/DeviceInfoStartAction.kt
package com.emarsys.mobileengage.device

import android.app.Activity
import com.emarsys.core.activity.ActivityLifecycleAction
import com.emarsys.core.activity.ActivityLifecycleAction.ActivityLifecycle.RESUME
import com.emarsys.core.activity.ActivityLifecyclePriorities
import com.emarsys.core.api.proxyApi
import com.emarsys.core.device.DeviceInfo
import com.emarsys.core.storage.Storage
import com.emarsys.mobileengage.client.ClientServiceInternal
import com.emarsys.mobileengage.di.mobileEngage

class DeviceInfoStartAction(private val clientInternal: ClientServiceInternal,
                            private val deviceInfoPayloadStorage: Storage<String?>,
                            private val deviceInfo: DeviceInfo,
                            override val triggeringLifecycle: ActivityLifecycleAction.ActivityLifecycle = RESUME,
                            override val priority: Int = ActivityLifecyclePriorities.DEVICE_INFO_START_ACTION_PRIORITY,
                            override val repeatable: Boolean = true
) : ActivityLifecycleAction {

    override fun execute(activity: Activity?) {
        val coreSdkHandler = mobileEngage().concurrentHandlerHolder
        coreSdkHandler.coreHandler.post {
            if (deviceInfoPayloadStorage.get() == null || deviceInfoPayloadStorage.get() != deviceInfo.deviceInfoPayload) {
                clientInternal.proxyApi(coreSdkHandler).trackDeviceInfo(null)
            }
        }
    }
}